### ISSUE TITLE
Ineffassign False Positive When Using Defer

### ISSUE THREAD

**Issue by coby-spotim** [golangci/golangci-lint#627]
1. Version of golangci-lint: v1.17.1-8-gd2b1eea
2. Config file: None
3. Go environment:
```
go version go1.12.6 darwin/amd64
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/coby/Library/Caches/go-build"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOOS="darwin"
GOPATH="/Users/coby/go"
GOPROXY=""
GORACE=""
GOROOT="/usr/local/Cellar/go/1.12.6/libexec"
GOTMPDIR=""
GOTOOLDIR="/usr/local/Cellar/go/1.12.6/libexec/pkg/tool/darwin_amd64"
GCCGO="gccgo"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/81/ylh1t84d2wjgb86zg857pxs00000gn/T/go-build162636616=/tmp/go-build -gno-record-gcc-switches -fno-common"
```
4. Verbose output of running: 
```
INFO [config_reader] Config search paths: [./ /Users/coby/Downloads /Users/coby /Users /]
INFO [lintersdb] Active 10 linters: [deadcode errcheck gosimple govet ineffassign staticcheck structcheck typecheck unused varcheck]
INFO [lintersdb] Optimized sublinters [staticcheck gosimple unused] into metalinter megacheck
INFO [loader] Go packages loading at mode load types and syntax took 596.152772ms
INFO [loader] SSA repr building timing: packages building 6.220155ms, total 35.986295ms
INFO [runner] worker.3 took 249.067µs with stages: errcheck: 180.449µs, deadcode: 34.781µs, varcheck: 16.99µs, structcheck: 6.962µs, typecheck: 3.425µs
INFO [runner] worker.4 took 751.89µs with stages: ineffassign: 747.962µs
ineffassign_check.go:10:3: ineffectual assignment to `ok` (ineffassign)
		ok = false
		^
INFO [runner] worker.1 took 122.957268ms with stages: govet: 122.949331ms
INFO [runner] worker.2 took 293.286229ms with stages: megacheck: 293.277879ms
INFO [runner] Workers idle times: #1: 170.30527ms, #3: 292.807582ms, #4: 292.393432ms
INFO [runner] processing took 822.767µs with stages: identifier_marker: 542.673µs, source_code: 93.892µs, exclude: 55.12µs, skip_dirs: 37.694µs, nolint: 35.33µs, autogenerated_exclude: 23.991µs, cgo: 18.475µs, path_prettifier: 4.946µs, max_same_issues: 3.015µs, path_shortener: 1.635µs, uniq_by_line: 1.559µs, max_from_linter: 1.338µs, filename_unadjuster: 906ns, max_per_file_from_linter: 773ns, diff: 677ns, skip_files: 459ns, exclude-rules: 284ns
INFO File cache stats: 1 entries of total size 205B
INFO Memory: 12 samples, avg is 147.1MB, max is 270.6MB
INFO Execution took 1.027496517s
```

Example Code Used for Test:
```go
package main

import "fmt"

func main() {
	ok := true
	defer fmt.Println(ok)
	err := someOperation()
	if err != nil {
		ok = false
		return
	}
}

func someOperation() error {
	return fmt.Errorf("Oh no!")
}
```

**Comment by coby-spotim** [comment]
I understand that since I'm declaring this in the defer, that the defer will execute with the original value of `ok` and not the adjusted value. I would expect to see a linting error telling me that I updated the `ok` value later in the code instead of telling me that it is an `ineffassign`.

**Comment by Helcaraxan** [comment]
My 2 cents: I believe the linter correctly identifies this as an `ineffassign`. You assign to the `ok` variable at line 10 (`ok = false`) without ever using the assigned value. This is the exact definition of an `ineffassign` error.

What I believe you are saying (correct me if I'm wrong), is that you would like a linter (not sure which one) to tell you that you are modifying a variable of which you have already captured the value in a `defer` invocation (the same would likely then hold for a value captured by a goroutine instantiation).

If that's indeed the case I'd offer the following slightly modified version of your example.
```go
package main

import "fmt"

func main() {
	val := 1
	defer fmt.Println(val)
	val = 2
	defer fmt.Println(val)
	val = 3
	fmt.Println(val)
}
```
What would the suggested linter check print in this case (if it prints anything at all)? Considering that the `ineffassign` linter won't print anything and the code is perfectly valid (although of debatable structure and quality 😛).

**Comment by coby-spotim** [comment]
I totally see your point, but my thought process here is that the linter should throw a warning anyways.
My understanding of linters is that they should tell you both if there are errors in your code, if the code is badly written according to the style/rules of the language, or if there is something suspicious about the code itself.
Even though the example that you wrote is valid go code, that doesn't mean that the code is right...

On the other hand I'm fairly new to Go, only been writing it a few months, so I might be out of my depth here... What do you think?

**Comment by Helcaraxan** [comment]
No, definitely not out of your depth. 🙂 

My response was in light of the title of this issue which stated that the reported linting error is a false positive, which given your response we can both agree it's not.

As I stated as well my code snipped is of debatable quality and there's definitely good arguments to be made to still have some form of static analysis that would disagree with it. If you are indeed new to Go then spotting this kind of valid-but-questionable code is proof of your understanding of the subtleties of the language.And I have to say that I do have some liking for the kind of check that you are suggesting, it's a good one! 🎉

Pragmatically speaking `golangci-lint` is a meta-linter and does not implement any linting or static analysis checks of its own. It rather relies on checks that it gets from other linters (such as `govet`, `staticcheck`, `ineffassign`, etc). As a result your "new check" suggestion would be more actionable if made to another GitHub project. The nature of the suggested check makes it, in my opinion, a prime candidate for [go-critic](https://github.com/go-critic/go-critic).

**Comment by jirfag** [comment]
totally agree with @Helcaraxan , closing
